### PR TITLE
Abundance timestep controls

### LINKED ISSUE #321: Consolidate timestep controls for changes in abundances of a given species
The timestep controls currently offer the following options, which are all particular cases of the same thing:
```f90
    dH_limit_min_H = 1d99
    dH_limit = 1d99
    dH_hard_limit = 1d99
    dH_decreases_only = .true.
    dH_div_H_limit_min_H = 1d-3
    dH_div_H_limit = 0.9d0
    dH_div_H_hard_limit = 1d99

    dHe_limit_min_He = 1d99
    dHe_limit = 1d99
    dHe_hard_limit = 1d99
    dHe_decreases_only = .true.
    dHe_div_He_limit_min_He = 1d-3
    dHe_div_He_limit = 0.9d0
    dHe_div_He_hard_limit = 1d99

    dHe3_limit_min_He3 = 1d99
    dHe3_limit = 1d99
    dHe3_hard_limit = 1d99
    dHe3_decreases_only = .true.
    dHe3_div_He3_limit_min_He3 = 1d99
    dHe3_div_He3_limit = 1d99
    dHe3_div_He3_hard_limit = 1d99

    dX_limit_min_X = 1d99
    dX_limit = 1d99
    dX_hard_limit = 1d99
    dX_decreases_only = .true.
    dX_div_X_limit_min_X = 1d99
    dX_div_X_limit = 1d99
    dX_div_X_hard_limit = 1d99
```
First, the `X` in the last set of controls refers to any non-H, non-He abundance (i.e., metals), which conflicts with the usual notation `Z`. But these controls could all be consolidated into arrays with a scheme something like:
```f90
    dX_limit_species(1) = 'h1' ! or any valid species
    dX_limit_min_X(1) = 1d99
    dX_limit(1) = 1d99
    dX_hard_limit(1) = 1d99
    dX_decreases_only(1) = .true.
    dX_div_X_limit_min_X(1) = 1d-3
    dX_div_X_limit(1) = 0.9d0
    dX_div_X_hard_limit(1) = 1d99
```
We could allow the special cases for species `'X'`, `'Y'` and `'Z'` for all H, all He and everything else (assuming that no-one will ever want to limit the timestep by changes in the Yttrium abundance).

### PR DESCRIPTION
**Pull request by warrickball**
Continuation of #481 (which should close #321) on MESAHub to test @matthiasfabry's changes before they're merged. Affected tests include `1M_thermohaline` and `c13_pocket`.

### CODE REVIEW

**Comment by warrickball** [comment]
We should log the changes before merging into `main`.

**Comment by warrickball** [comment]
A bunch of tests [failed](https://testhub.mesastar.org/abundance_timestep_controls/commits/67ca7d9) because their inlists need to be updated to the new controls.

**Comment by warrickball** [comment]
Oh... This has changed some defaults with subtle consequences in some tests.

On investigating some of the failures, I noticed that we previously had some non-trivial defaults that would (quite loosely) limit how quickly a cell can deplete H or He. Here are some relevant defaults from [the current release](https://docs.mesastar.org/en/release-r22.11.1/reference/controls.html#dh-decreases-only):
```f90
    dH_decreases_only = .true.
    dH_div_H_limit_min_H = 1d-3
    dH_div_H_limit = 0.9d0
    
    dHe_decreases_only = .true.
    dHe_div_He_limit_min_He = 1d-3
    dHe_div_He_limit = 0.9d0
```
One quiet consequence here is that `1M_pre_ms_to_wd` goes from taking 13887 steps [here](https://testhub.mesastar.org/abundance_timestep_controls/commits/112459a/test_cases/star/1M_pre_ms_to_wd) to taking 5032 [here](https://testhub.mesastar.org/abundance_timestep_controls/commits/67ca7d9/test_cases/star/1M_pre_ms_to_wd). (Focus on the `bluebear_ifort` numbers, which are for full runs.) I presume that a large part of this is because the star can race through shell burning, particularly on the (first-ascent) RGB, by completely depleting cells in a single step.

So besides fixing the inlists of the failing tests, we must either
1. restore the old default behaviour or
2. re-assess the relevant controls in individual tests whose behaviour is changed.

I'm personally in favour of 1 and am not sure there's any strong case for 2. This is the sort of thing that I suspect is a hard-learned lesson after years of getting tests and other runs to converge. I think it boils down to
```f90
    dX_limit_species(1) = 'h1' ! or 'X'? not sure which reproduces old dH control
    dX_decreases_only(1) = .true.
    dX_div_X_limit_min_X(1) = 1d-3
    dX_div_X_limit(1) = 0.9d0

    dX_limit_species(2) = 'he4' ! or 'Y'? not sure which reproduces old dHe control
    dX_decreases_only(2) = .true.
    dX_div_X_limit_min_X(2) = 1d-3
    dX_div_X_limit(2) = 0.9d0
```
This can also be written a bit more tightly in array notation but we have to be careful of confusing users because Fortran source (which is how the defaults are included) and Fortran namelists behave slightly differently.

I think this is going to make the docs look weird but I'm happy to tackle that later. I think this is a good change so want to get the test suite back on track so we can finally merge it.

**Comment by fxt44** [comment]
super! test suite rolling on 632539e.

**Review by evbauer — APPROVED** [review]

**Comment by warrickball** [comment]
Changing the defaults hasn't restored the old behaviour. I'm just looking at `1M_pre_ms_to_wd`, which is going up the RGB much faster, as can be seen by comparing the `to_start_he_core_flash` tab in the output [before](https://testhub.mesastar.org/abundance_timestep_controls/commits/a130cfd/test_cases/star/1M_pre_ms_to_wd) or [after](https://testhub.mesastar.org/abundance_timestep_controls/commits/632539e/test_cases/star/1M_pre_ms_to_wd) these changes (including my changes to the defaults). I don't think my changes accomplished anything...

File: star/private/timestep.f90 [hunk 1]
```diff
@@ -495,226 +480,192 @@ integer function check_diffusion_iters_limit(s, skip_hard_limit, dt, dt_limit_ra
...
+         do i=1, max_dx_limit_ctrls  ! go over all potential species + XYZ
+            if (  dX_limit_min_X(i) >= 1 .or. &  ! as soon as any of these is >= 1
+                  dX_limit(i) >= 1 .or. &        ! we'd have a problem elsewhere
+                  dX_hard_limit(i) >= 1 .or. &
+                  dX_div_X_limit_min_X(i) >= 1 .or. &
+                  dX_div_X_limit(i) >= 1 .or. &
+                  dX_div_X_hard_limit(i) >= 1 .or. &
+                  s% dX_limit_species(i) == '') then
+               cycle  ! go to next
             end if
```
[the remark this answers is not in the document]
Comment (warrickball) [reply star/private/timestep.f90:525]: Also, my apologies: I made this comment a week or two ago but apparently it didn't turn up for anyone else until I marked my "review" as finished.

[the remark this answers is not in the document]
Comment (matthiasfabry) [reply star/private/timestep.f90:525]: Of course, good catch. I've changed the behavior on my forked branch, so please merge that to start testing. I don't think I can directly push on this one yet.

**Comment by evbauer** [comment]
Well, I was a bit clumsy and ended up creating more merge commits than necessary, but the upshot of all that should be just that we got the updated logic from [2f03229](https://github.com/MESAHub/mesa/pull/484/commits/2f03229b9ce1b835b9341b445762308555933c17) in here now.

**Comment by matthiasfabry** [comment]
That's entirely my bad, Evan. I merged main into my branch after making the improvements, which I guess is bad practice? I can revert this easily on my local branch, but I couldn't quickly find a way to revert this upstream. I guess I need command line stuff for this, which I'm unfamiliar with at this point.

**Comment by evbauer** [comment]
No worries. I wouldn't have thought merging like you did would be an issue. Somehow the fact that I also did a merge before trying to do a PR from your fork really confused the web interface, which thought there were merge conflicts. So it told me to go through the command line to resolve merge conflicts, but then on the command line there weren't any actual conflicts to resolve. The net result is just a silly series of merge commits with no actual content, but I think we've ended up where we need to be regardless.

**Comment by warrickball** [comment]
Is the behaviour back to what it was before we changed the control structure? My thinking is that these changes should be purely cosmetic; no results should change, as long as the same default are implemented as before (which is how we noticed we were skipping things).

It looks like we don't have any full runs on the most recent commit so I'll submit a run to our cluster now and we can then compare the before and after output.

I'll use [test results for 152c8fd](https://testhub.mesastar.org/abundance_timestep_controls/commits/152c8fd) as a reference, i.e. "before".

**Comment by fxt44** [comment]
i can run the test suite - what's the best version to run?

**Comment by warrickball** [comment]
ca08c78 is the commit at the top of this branch.

**Comment by warrickball** [comment]
In ca08c78, `1M_pre_ms_to_wd` is still going up the RGB much faster than it did before: ~2000 steps rather than ~10000 like is used to (see the `to_start_he_core_flash` tab [here](https://testhub.mesastar.org/abundance_timestep_controls/commits/ca08c78/test_cases/star/1M_pre_ms_to_wd)). That still suggests to me that the abundance control, which is what used to limit the timestep, is not functioning as it did before and therefore needs investigating. I'm happy to take a look but it might take me ~10 days to do so.

**Comment by matthiasfabry** [comment]
ok let me see if I can figure it out

**Comment by fxt44** [comment]
rolling two machines on the 2e2dc1f test suite ...

**Comment by warrickball** [comment]
I did a full run on `bluebear_ifort` and [it looks like](https://testhub.mesastar.org/abundance_timestep_controls/commits/2e2dc1f/test_cases/star/1M_pre_ms_to_wd) `1M_pre_ms_to_wd` is still rocketing up the RGB. You can try to simply run it locally, I doubt it takes more than a few minutes to start going up the RGB and hopefully for you to see it finish ascending after ~2000 steps.

Alternatively, if you put `[ci optional]` in the commit message, I think `cannon_conroy_intel` will automatically do a full run, though @evbauer can correct me if `cannon_conroy_intel` only checks for `[ci optional]` on `main`. If that works, you can just wait for the output on the TestHub.

If you aren't familiar with TestHub, [this is the link](https://testhub.mesastar.org/abundance_timestep_controls/commits) to the tests of commits along this branch. You can check there for whether `1M_pre_ms_to_wd` does (not) go back to taking ~10000 steps on the RGB.

**Comment by matthiasfabry** [comment]
`cannon_conroy_intel` [now](https://testhub.mesastar.org/abundance_timestep_controls/commits/d553177/test_cases/star/1M_pre_ms_to_wd) gives ~10000 steps again for the RBG climb! Does this require further tests on `ifort` or other architectures @warrickball?

**Comment by warrickball** [comment]
That looks better. Thanks! I've submitted a run on `bluebear_ifort`, just to be sure, but if that all passes (except for the known issue with `rsp_save_and_load_file`), I think this is okay to merge. If any other devs want to chip in, now is probably the time.

Out of curiosity, what turned out to be the issue?

**Comment by matthiasfabry** [comment]
It was that `check_dX` is now just one call for all species, rather than several previously. I was overwriting the `limit_ratio` of H (which should trigger) with that one of He (which doesn't) in the next pass of the species loop within `check_dX`, so the timestep controller saw nothing wrong.

**Comment by warrickball** [comment]
Looks good. Here's [another view](https://testhub.mesastar.org/abundance_timestep_controls/test_cases/star/1M_pre_ms_to_wd?utf8=%E2%9C%93&history_type=show_instances&start_date=2022-11-01&end_date=2023-02-18&computers=bluebear_ifort&status=&commit=Search) of `1M_pre_ms_to_wd` on `bluebear_ifort`, which always does complete runs. It's not trivial to make sense of the commits but it's clear that the behaviour is now back to what it was on `main` a few months ago (including around when r22.11.1 was released).

**Comment by evbauer** [comment]
I have a minor concern that this update could cause us to lose some information from terminal output about which specific species is causing the abundance change that limits the timestep. If there's an easy way to update to include that information in the terminal output, then I think that would be worth trying.

Otherwise, it looks like this is ready to merge. Let's shoot for merging it within the next week or so unless there are objections.

File: star/private/timestep.f90 [hunk 2]
```diff
@@ -495,226 +480,203 @@ integer function check_diffusion_iters_limit(s, skip_hard_limit, dt, dt_limit_ra
...
-                  max_dX_div_X_bdy_dist_dm/Msun
+   
+            if (dX_div_X_limit(i) > 0) then
+               ratio_tmp = max_dX_div_X/dX_div_X_limit(i)
+               if (ratio_tmp > dX_div_X_dt_limit_ratio) then  ! pick out largest culprit only!
+                  dX_div_X_dt_limit_ratio = ratio_tmp
+                  if (dX_div_X_dt_limit_ratio <= 1d0) then
+                     dX_div_X_dt_limit_ratio = 0
+                  else
+                     s% Tlim_dX_div_X_species = max_dX_div_X_j
```
Comment (matthiasfabry) [inline star/private/timestep.f90:667]: @evbauer this property of star is still set, so it _should_ be possible to map that back to the species. As stated though, I think the printing of the string should happen elsewhere, not in this routine.

**Comment by evbauer** [comment]
There was a small initialization problem for the `ratio_tmp` variable, so I've pushed a fix for that. Previously, `1M_pre_ms_to_wd` was reporting the timestep limit as just `dX` in the terminal output for the many steps ascending the RGB. Now it reports `dX/X h1`, so I think we're in good shape!

**Pull request merged**